### PR TITLE
fix(gatsby): Load static query results from its own Db

### PR DESCRIPTION
**Pull request by sidharthachatterjee**

### CODE REVIEW

File: packages/gatsby/cache-dir/loader.js [hunk 1]
```diff
@@ -529,6 +529,7 @@ export const publicLoader = {
   isPageNotFound: rawPath => instance.isPageNotFound(rawPath),
   hovering: rawPath => instance.hovering(rawPath),
   loadAppData: () => instance.loadAppData(),
+  getStaticQueryDb: () => Object.fromEntries(instance.staticQueryDb),
```
Comment (pieh) [inline packages/gatsby/cache-dir/loader.js:532]: Can we instead just have `staticQueryDb` as plain object and not Map to save on those conversions on each render? Also `Object.fromEntries` would need polyfilling possibly? It's quite new API

Comment (sidharthachatterjee) [reply packages/gatsby/cache-dir/loader.js:532]: Yeah, we'll need a polyfill. I agree that a plain object is probably a better idea. Map is better semantically and more consistent with the rest of the dbs though.

Comment (sidharthachatterjee) [reply packages/gatsby/cache-dir/loader.js:532]: Done

**Comment by sidharthachatterjee** [comment]
Just published a canary at `gatsby@static-query-fix`

File: packages/gatsby/cache-dir/production-app.js [hunk 2]
```diff
@@ -70,15 +70,16 @@ apiRunnerAsync(`onClientEntry`).then(() => {
...
-                >
-                  <DataContext.Provider value={{ pageResources, location }}>
-                    {children}
-                  </DataContext.Provider>
-                </StaticQueryContext.Provider>
-              )}
+              {({ pageResources, location }) => {
+                const staticQueryResults = publicLoader.getStaticQueryResults()
+                return (
+                  <StaticQueryContext.Provider value={staticQueryResults}>
```
Comment (sidharthachatterjee) [inline packages/gatsby/cache-dir/production-app.js:76]: This keeps the `StaticQueryContext` public API and also should be a lot more stable (fixing cases like we saw for `gatsby-plugin-transition-link`)

File: packages/gatsby/cache-dir/loader.js (hunk 1 above)
Comment (pvdz) [reply packages/gatsby/cache-dir/loader.js]: Might have an adverse perf implication if there are many of these queries. I don't know if that number scales with the number of pages or just the number of source code files.

Comment (pvdz) [reply packages/gatsby/cache-dir/loader.js]: Additionally, the returned object is now _shared_ by reference, where it was a fresh reference before. Not sure how relevant that is..?

Comment (pieh) [reply packages/gatsby/cache-dir/loader.js]: > I don't know if that number scales with the number of pages or just the number of source code files.

This scales with number of source files, not nodes/data

Comment (sidharthachatterjee) [reply packages/gatsby/cache-dir/loader.js]: @pvdz Great questions! 

Shared by reference is by design and should be fine. And as @pieh mentioned, `n` here will be the number of unique static queries and templates (and not pages built from these templates) so we should be fine.

**Comment by sidharthachatterjee** [comment]
Published in `gatsby@2.24.13`

**Comment by pvdz** [comment]
> Let's merge this because it fixes the bug. If swap from Map to plain object cause actual perf issues we can iterate

And how are you going to find out? :see_no_evil:

**Comment by sidharthachatterjee** [comment]
@pvdz That's a very good question... 😛

**Comment by r3sMetz** [comment]
Maybe unrelated, but i still got the issue when running useStaticQuery inside /src/stories (Storybook)... Worth an issue?

**Comment by sidharthachatterjee** [comment]
@r3sMetz Yes, storybook is currently broken and a fix is coming soon. Feel free to open an issue and assign to me! (There might already be one)

**Pull request merged**